File: 875_Koko Eating Bananas.py
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        minK = 1
        maxK = max(piles)
        
        res = maxK

        while minK <= maxK:
            mid = (minK+maxK)//2
            hours =0
            for i in piles:
                hours += math.ceil(i/mid)

            if hours <= h:
                res = min(res, mid)
                maxK = mid -1
            else:
                minK = mid+1

        return res

            
        return self.searchHelper(1, maxK, h, piles)


    def searchHelper(self, low: int, high: int, h: int, piles: List[int]) -> int:
        mid = (low+high)//2
        hours = self.hoursTaken(mid, piles)
        if hours <= h:
            for i in range(mid, low-1, -1):
                if self.hoursTaken(i, piles) > h:
                    return i+1
                
        if hours > h:
            return self.searchHelper(mid+1, high, h, piles)


    def hoursTaken(self, k: int, piles: List[int]) -> int:
        sumK = 0
        for i in piles:
            if not (i % k==0):
                sumK+=1
            sumK += (i//k)
        
        return sumK
    

s = Solution()
print(f"final is {s.minEatingSpeed([312884470], 312884469)}")

File: Java Leetcode/875_Koko Eating Bananas.java
import java.util.*;

class Solution {

    public static void main(String[] args) {
        System.out.println(3 % 4 == 0);
    }

    public int minEatingSpeed(int[] piles, int h) {
        int minK = 1;
        int maxK = Arrays.stream(piles).max().orElse(0);

        if (piles.length == 1) {
            return piles[0] / h;
        }

        if (h == piles.length) {
            return maxK;
        }

        if (h > Arrays.stream(piles).sum()) {
            return 1;
        }

        return searchHelper(1, maxK, h, piles);
    }

    private int searchHelper(int low, int high, int h, int[] piles) {
        int mid = (low + high) / 2;
        int hours = hoursTaken(mid, piles);

        if (hours == h) {
            for (int i = low; i <= mid; i++) {
                if (hoursTaken(i, piles) == h) {
                    return i;
                }
            }
        }

        if (hours > h) {
            return searchHelper(mid + 1, high, h, piles);
        }

        if (hours < h) {
            return searchHelper(1, mid - 1, h, piles);
        }

        return -1; // or handle other cases based on your requirements
    }

    private int hoursTaken(int k, int[] piles) {
        int sumK = 0;
        for (int i : piles) {
            if (!(i % k == 0)) {
                sumK += 1;
            }
            sumK += (i / k);
        }

        return sumK;
    }
}
